lib/ClimaCoreSpectra/src/gcm_spectra.jl

```diff
@@ -0,0 +1,392 @@
+
+# Arguments
+- FT: FloatType
+- n: number of Gaussian latitudes
+
+# References
+- Ehrendorfer, M., Spectral Numerical Weather Prediction Models, Appendix B, Society for Industrial and Applied Mathematics, 2011
+# Details (following notation from Ehrendorfer, 2011):
+    Pn(x) is an odd function
+    solve half of the n roots and weightes of Pn(x) # n = 2n_half
+    P_{-1}(x) = 0
+    P_0(x) = 1
+    P_1(x) = x
+    nP_n(x) = (2n-1)xP_{n-1}(x) - (n-1)P_{n-2}(x)
+    P'_n(x) = n/(x^2-1)(xP_{n}(x) - P_{n-1}(x))
+    x -= P_n(x)/P'_{n}()
+    Initial guess xi^{0} = cos(π(i-0.25)/(n+0.5))
+    wi = 2/(1-xi^2)/P_n'(xi)^2
+"""
+function compute_gaussian!(FT, n)
+    itermax = 10000
+    tol = 1.0e-15
+
+    sinθ = zeros(FT, n)
+    wts = zeros(FT, n)
+
+    n_half = Int(n / 2)
+    for i in 1:n_half
+        dp = 0.0
+        z = cos(pi * (i - 0.25) / (n + 0.5))
+        for iter in 1:itermax
+            p2 = 0.0
+            p1 = 1.0
+
+            for j in 1:n
+                p3 = p2 # Pj-2
+                p2 = p1 # Pj-1
+                p1 = ((2.0 * j - 1.0) * z * p2 - (j - 1.0) * p3) / j  #Pj
+            end
+            # P'_n
+            dp = n * (z * p1 - p2) / (z * z - 1.0)
+            z1 = z
+            z = z1 - p1 / dp
+            if (abs(z - z1) <= tol)
+                break
+            end
+            if iter == itermax
+                @error("Compute_Gaussian! does not converge!")
+            end
+        end
+
+        sinθ[i], sinθ[n - i + 1], = -z, z
+        wts[i] = wts[n - i + 1] = 2.0 / ((1.0 - z * z) * dp * dp)
+    end
+
+    return sinθ, wts
+end
+
+"""
+    trans_grid_to_spherical!(mesh::SpectralSphericalMesh, pfield::Arr{FT,2})
+
+Transforms a variable on a Gaussian grid (pfield[nλ, nθ]) into the spherical harmonics domain (var_spherical2d[num_fourier+1, num_spherical+1])
+Here λ = longitude, θ = latitude, η = sinθ, m = zonal wavenumber, n = total wavenumber:
+var_spherical2d = F_{m,n}    # Output variable in spectral space (Complex{FT}[num_fourier+1, num_spherical+1])
+qwg = P_{m,n}(η)w(η)         # Weighted Legendre polynomials (FT[num_fourier+1, num_spherical+1, nθ])
+var_fourier2d = g_{m, θ}     # Untruncated Fourier transformation (Complex{FT} [nλ, nθ])
+pfield = F(λ, η)             # Input variable on Gaussian grid FT[nλ, nθ]
+
+# Arguments
+- mesh: struct with mesh information
+- pfield: variable on Gaussian grid to be transformed
+
+# References
+- Ehrendorfer, M., Spectral Numerical Weather Prediction Models, Appendix B, Society for Industrial and Applied Mathematics, 2011
+- [Wiin1967](@cite)
+"""
+function trans_grid_to_spherical!(
+    mesh::SpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4},
+    pfield::ArrF2,
+) where {FT, ArrF3, ArrI2, ArrC3, ArrC4, ArrF2}
+
+    num_fourier, num_spherical = mesh.num_fourier, mesh.num_spherical
+    var_fourier2d, var_spherical2d =
+        mesh.var_fourier[:, :, 1] * 0.0, mesh.var_spherical[:, :, 1, :] * 0.0
```

Comment:
```suggestion
        mesh.var_fourier[:, :, 1] * 0, mesh.var_spherical[:, :, 1, :] * 0
```

Reply:
It looks like you're timing in global scope. Please see https://docs.julialang.org/en/v1/manual/performance-tips/#Performance-critical-code-should-be-inside-a-function

Reply:
Thanks for the reminder. It seems like the change would not provide benefits then:
```
julia> function with_float() 
       var_fourier2d, var_spherical2d =
                      mesh.var_fourier[:, :, 1] * 0.0, mesh.var_spherical[:, :, 1, :] * 0.0
       end
with_float (generic function with 1 method)

julia> function with_int() 
       var_fourier2d, var_spherical2d =
                      mesh.var_fourier[:, :, 1] * 0, mesh.var_spherical[:, :, 1, :] * 0
       end
with_int (generic function with 1 method)

julia> @time with_float()
  0.000081 seconds (42 allocations: 318.094 KiB)
  
julia> @time with_int()
  0.000080 seconds (42 allocations: 318.094 KiB)
```
I'll be happy to discuss more offline

Reply:
Let's make the changes I suggested and we can discuss the performance issues offline.